File: dp-lab3/c/parrot.c
#include <stdlib.h>

typedef char const* (*PTRFUN)();
typedef struct Parrot Parrot;

char const* name(Parrot* parrot);
char const* greet();
char const* menu();

void* create(char const* name);
void construct(char const* name, Parrot* memorySpace);
size_t size_of();

struct Parrot {
    PTRFUN *funTable;
    const char *name;
};

PTRFUN vtable[] = {name, greet, menu};

char const* name(Parrot* parrot) {
    return parrot->name;
}

char const* greet() {
    return "squawk!";
}

char const* menu() {
    return "sjemenke";
}

void* create(char const* name) {
    Parrot* parrot = (Parrot*)malloc(sizeof(Parrot));
    construct(name, parrot);
    return parrot;
}

void construct(char const* name, Parrot* parrot) {
    parrot->funTable = vtable;
    parrot->name = name;
}

size_t size_of() {
    return sizeof(Parrot);
}

File: dp-lab3/java/src/hr/fer/zemris/ooup/lab3/model/plugins/Parrot.java
package hr.fer.zemris.ooup.lab3.model.plugins;

import hr.fer.zemris.ooup.lab3.model.Animal;

public class Parrot extends Animal {
	
	private String animalName;
	
	public Parrot(String name) {
		this.animalName = name;
	}

	@Override
	public String name() {
		return animalName;
	}

	@Override
	public String greet() {
		return "squawk!";
	}

	@Override
	public String menu() {
		return "sjemenke";
	}
	
}
